### PR TITLE
[InstCombine] Canonicalise SextADD + GEP

### PR DESCRIPTION
**Pull request by LiqinWeng**
For a two-dimensional array, the base address does not need to be recalculated for the second access. It can be obtained through offset. I will submit it later: fold (mul (sext (add_nsw x, c1)), c2) -> (add (mul (sext x), c2), c1*c2) 
The code scenario is as follows
```
typedef int array1[50];
void test(int array2[50][50], int a, int b) {
    int loc;
    loc = a + 5;
    array2[loc][loc] += 1;
    array2[loc + 20][loc] = loc;
}
```
form dhrystone.

### CODE REVIEW

**Comment by goldsteinn** [comment]
Can you add alive2 proof?

**Comment by LiqinWeng** [comment]
prepare for https://github.com/llvm/llvm-project/pull/69667

**Comment by LiqinWeng** [comment]
> Can you add alive2 proof?

Alive2: https://alive2.llvm.org/ce/z/VYoq5-

**Review by nikic — COMMENTED** [review]
It's not at all clear to me that this is a generally profitable transform. Just based on the test changes here, loopflatten.ll is an obvious regression (loop no longer flattened), while induction.ll also looks like a regression to me.

This is also a regression on the simple instruction count heuristic, as we go from add+sext+gep to 2*sext+2*gep. Possibly what you want to do is limit this to the case where the RHS of the add is a constant, so the sext folds away?

**Comment by LiqinWeng** [comment]
> It's not at all clear to me that this is a generally profitable transform. Just based on the test changes here, loopflatten.ll is an obvious regression (loop no longer flattened), while induction.ll also looks like a regression to me.
>  
> This is also a regression on the simple instruction count heuristic, as we go from add+sext+gep to 2_sext+2_gep. Possibly what you want to do is limit this to the case where the RHS of the add is a constant, so the sext folds away?

```
void test(int array2[50][50], int a, int b) {
    int loc;
    loc = a + 5;
    array2[loc][loc] += 1;
    array2[loc + 20][loc] = loc;
}
```
For a two-dimensional array, the base address does not need to be recalculated for the second access. It can be obtained through offset.

**Comment by LiqinWeng** [comment]
> This is also a regression on the simple instruction count heuristic, as we go from add+sext+gep to 2_sext+2_gep. Possibly what you want to do is limit this to the case where the RHS of the add is a constant, so the sext folds away?
yes, I have fixed it, thanks

**Comment by LiqinWeng** [comment]
> It's not at all clear to me that this is a generally profitable transform. Just based on the test changes here, loopflatten.ll is an obvious regression (loop no longer flattened), while induction.ll also looks like a regression to me.
> 
> This is also a regression on the simple instruction count heuristic, as we go from add+sext+gep to 2_sext+2_gep. Possibly what you want to do is limit this to the case where the RHS of the add is a constant, so the sext folds away?

yse, I have fix it , thanks

**Comment by nikic** [comment]
Please, can you take a bit more care? Your proof is very obviously broken if you just look at the result.

Here is a fixed version: https://alive2.llvm.org/ce/z/FSNnoo Note that the `nsw` on the `add` is required for correctness, which you don't check.

**Comment by LiqinWeng** [comment]
> Please, can you take a bit more care? Your proof is very obviously broken if you just look at the result.
> 
> Here is a fixed version: https://alive2.llvm.org/ce/z/FSNnoo Note that the `nsw` on the `add` is required for correctness, which you don't check.

Pls see: https://alive2.llvm.org/ce/z/vkCK92. The transformation only occurs when the second operand of the add is a constant.
I will add the `nsw` on the `add`

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 1]
```diff
@@ -2335,6 +2335,27 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
...
+              m_OneUse(m_Add(m_Value(Idx1), m_ConstantInt(C)))) &&
+        hasNoSignedWrap(*AddOp)) {
+      // %add = add nsw i32 %idx, C
+      // %sidx = sext i32 %add to i64
+      // %gep = getelementptr i32, i32* %ptr, i64 %sidx
+      // as :
+      // %sidx = sext i32 %idx to i64
+      // %newptr = getelementptr i32, i32* %ptr, i64 %sidx
+      // %newgep = getelementptr i32, i32* %newptr, i64 C
+      auto SIdx1 = Builder.CreateSExt(Idx1, GEP.getOperand(1)->getType());
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2354]: Do you even need the new sext?
https://alive2.llvm.org/ce/z/dgfX5w

**Comment by goldsteinn** [comment]
> > Please, can you take a bit more care? Your proof is very obviously broken if you just look at the result.
> > Here is a fixed version: https://alive2.llvm.org/ce/z/FSNnoo Note that the `nsw` on the `add` is required for correctness, which you don't check.
> 
> Pls see: https://alive2.llvm.org/ce/z/vkCK92. The transformation only occurs when the second operand of the add is a constant. I will add the `nsw` on the `add`

The proof is only for constant 5. In general you should have no constants and impl any constraints with assume:
https://alive2.llvm.org/ce/z/dgfX5w

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp (hunk 1 above)
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2354]: Yes, in the sense that it would just get canonicalized to add the sext anyway.

Comment (LiqinWeng) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp]: the test: loopflatten.ll  is regression?

**Comment by LiqinWeng** [comment]
ping

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 2]
```diff
@@ -2335,6 +2335,28 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
...
                                        Idx2);
     }
+
+    Value *SIdx;
+    if (auto *SextInst = dyn_cast<SExtInst>(GEP.getOperand(1))) {
+      if (auto *AddInst = dyn_cast<BinaryOperator>(SextInst->getOperand(0))) {
+        if (match(GEP.getOperand(1), m_OneUse(m_SExt(m_Value(SIdx)))) &&
+            match(SextInst->getOperand(0),
+                  m_OneUse(m_Add(m_Value(Idx1), m_Value(Idx2)))) &&
+            hasNoSignedWrap(*AddInst)) {
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2345]: Should do something like `if (match(GEP.getOperand(1), m_OneUse(m_SExt(m_OneUse(m_NSWAdd(m_Value(Idx1), m_Value(Idx2))))))` here.

File: polly/test/Support/dumpmodule.ll [hunk 3]
```diff
@@ -1,3 +1,4 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 3
```
Comment (nikic) [inline polly/test/Support/dumpmodule.ll:1]: No they haven't. Drop this line and and the spurious changes to the test.

File: llvm/test/Transforms/PhaseOrdering/AArch64/loopflatten.ll [hunk 4]
```diff
@@ -10,20 +10,29 @@ define dso_local void @_Z3fooPiii(ptr %A, i32 %N, i32 %M) #0 {
...
+; CHECK-NEXT:    [[ARRAYIDX_US:%.*]] = getelementptr i32, ptr [[TMP2]], i64 [[INDVARS_IV]]
+; CHECK-NEXT:    [[TMP3:%.*]] = load i32, ptr [[ARRAYIDX_US]], align 4
+; CHECK-NEXT:    tail call void @_Z1fi(i32 [[TMP3]])
+; CHECK-NEXT:    [[INDVARS_IV_NEXT]] = add nuw nsw i64 [[INDVARS_IV]], 1
+; CHECK-NEXT:    [[EXITCOND_NOT:%.*]] = icmp eq i64 [[INDVARS_IV_NEXT]], [[WIDE_TRIP_COUNT]]
+; CHECK-NEXT:    br i1 [[EXITCOND_NOT]], label [[FOR_COND1_FOR_COND_CLEANUP3_CRIT_EDGE_US]], label [[FOR_BODY4_US]]
+; CHECK:       for.cond1.for.cond.cleanup3_crit_edge.us:
+; CHECK-NEXT:    [[INDVARS_IV_NEXT8]] = add nuw nsw i64 [[INDVARS_IV7]], 1
+; CHECK-NEXT:    [[EXITCOND12_NOT:%.*]] = icmp eq i64 [[INDVARS_IV_NEXT8]], [[WIDE_TRIP_COUNT11]]
+; CHECK-NEXT:    br i1 [[EXITCOND12_NOT]], label [[FOR_COND_CLEANUP]], label [[FOR_COND1_PREHEADER_US]]
```
Comment (nikic) [inline llvm/test/Transforms/PhaseOrdering/AArch64/loopflatten.ll:35]: This is still a regression.

**Comment by nikic** [comment]
Also, why did you drop the limitation to having one constant operand again?

File: llvm/test/Transforms/PhaseOrdering/AArch64/loopflatten.ll (hunk 4 above)
Comment (LiqinWeng) [reply llvm/test/Transforms/PhaseOrdering/AArch64/loopflatten.ll]: fixed

File: polly/test/Support/dumpmodule.ll (hunk 3 above)
Comment (LiqinWeng) [reply polly/test/Support/dumpmodule.ll]: fixed

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp (hunk 2 above)
Comment (LiqinWeng) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp]: done

**Comment by LiqinWeng** [comment]
> Also, why did you drop the limitation to having one constant operand again?

Sorry，I misunderstood what you meant, and I have corrected it

**Comment by LiqinWeng** [comment]
> Also, why did you drop the limitation to having one constant operand again?

I have a question. Why is there no nuw flag when executing the following code on a 64-bit platform?

code：
`void test1(unsigned int array1[50], unsigned int a, unsigned int b) {
    unsigned int loc;
    loc = a + 5;
    array1[loc] = b;
    array1[loc + 1] = array1[loc];
    array1[loc + 30] = loc;
}`

**Comment by LiqinWeng** [comment]
> > Also, why did you drop the limitation to having one constant operand again?
> 

I have a question:  Why is there no nuw flag on test1, but test2 has nsw flag. 
command: clang --target=riscv64 shadd.c -O3 -emit-llvm -S -o -
code：
```
void test1(unsigned int array1[50], unsigned int a, unsigned int b) {
    unsigned int loc;
    loc = a + 5;
    array1[loc] = b;
    array1[loc + 1] = array1[loc];
    array1[loc + 30] = loc;
}

void test2(int array1[50], int a, int b) {
    int loc;
    loc = a + 5;

    array1[loc] = b;
    array1[loc + 1] = array1[loc];
    array1[loc + 30] = loc;
}
```

image

**Comment by LiqinWeng** [comment]
> > Also, why did you drop the limitation to having one constant operand again?
> 

I have a question:  Why is there no nuw flag on test1, but test2 has nsw flag. 
command: clang --target=riscv64 shadd.c -O3 -emit-llvm -S -o -
code：
```
void test1(unsigned int array1[50], unsigned int a, unsigned int b) {
    unsigned int loc;
    loc = a + 5;
    array1[loc] = b;
    array1[loc + 1] = array1[loc];
    array1[loc + 30] = loc;
}

void test2(int array1[50], int a, int b) {
    int loc;
    loc = a + 5;

    array1[loc] = b;
    array1[loc + 1] = array1[loc];
    array1[loc + 30] = loc;
}
```

image

**Comment by LiqinWeng** [comment]
> Also, why did you drop the limitation to having one constant operand again?
For this code scenario, if the constants are not extracted in the instcombine stage, the base address needs to be recalculated the second time the array is accessed. How transform to this scenario
code：
command: clang --target=riscv64 shadd.c -O3 -emit-llvm -S -o -
```
void test1(unsigned int array1[50], unsigned int a, unsigned int b) {
    unsigned int loc;
    loc = a + 5;
    array1[loc] = b;
    array1[loc + 1] = array1[loc];
    array1[loc + 30] = loc;
}
```

**Comment by LiqinWeng** [comment]
ping

**Comment by LiqinWeng** [comment]
Are there any issues with this patch? Can it be merge？
@nikic

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 5]
```diff
@@ -2335,6 +2335,20 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
...
+
+    ConstantInt *C;
+    if (match(GEP.getOperand(1), m_OneUse(m_SExt(m_OneUse(m_NSWAdd(
+                                     m_Value(Idx1), m_ConstantInt(C))))))) {
+      // %add = add nsw i32 %idx1, idx2
+      // %sidx = sext i32 %add to i64
+      // %gep = getelementptr i32, i32* %ptr, i64 %sidx
+      // as :
+      // %newptr = getelementptr i32, i32* %ptr, i32 %idx1
+      // %newgep = getelementptr i32, i32* %newptr, i32 idx2
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2347]:
```suggestion
      // %gep = getelementptr i32, ptr %ptr, i64 %sidx
      // as:
      // %newptr = getelementptr i32, ptr %ptr, i32 %idx1
      // %newgep = getelementptr i32, ptr %newptr, i32 idx2
```

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 6]
```diff
@@ -2335,6 +2335,20 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
...
+      auto *NewPtr = Builder.CreateGEP(GEP.getResultElementType(),
+                                       GEP.getPointerOperand(), Idx1);
+      return GetElementPtrInst::Create(GEP.getResultElementType(), NewPtr, C);
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2350]:
```suggestion
      auto *NewPtr = Builder.CreateGEP(GEP.getResultElementType(),
                                       GEP.getPointerOperand(), Builder.CreateSExt(Idx1));
      return GetElementPtrInst::Create(GEP.getResultElementType(), NewPtr, Builder.CreateSExt(C));
```
Make the sexts explicit please.

File: llvm/test/Transforms/InstCombine/array.ll [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  %0 = load i32, ptr %arrayidx2, align 4
+  %add3 = add nsw i32 %0, 1
+  store i32 %add3, ptr %arrayidx2, align 4
+  %add4 = add nsw i32 %add, 20
+  %idxprom5 = sext i32 %add4 to i64
+  %arrayidx6 = getelementptr inbounds [50 x i32], ptr %array2, i64 %idxprom5
+  %arrayidx8 = getelementptr inbounds [50 x i32], ptr %arrayidx6, i64 0, i64 %idxprom
+  store i32 %add, ptr %arrayidx8, align 4
+  ret void
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/array.ll:38]: Minimize this test to just the pattern being transformed, drop all the other irrelevant instructions.

You also need negative tests for all the preconditions of the transforms, e.g. the one-use checks, the nsw on the add, the constant operand.

File: polly/test/Support/dumpmodule.ll [hunk 8]
```diff
@@ -68,8 +68,7 @@ return:
 
 
 ; EARLY-LABEL: @callee(
-; AFTEREARLY:  polly.split_new_and_old:
-; EARLY:         store double 4.200000e+01, ptr %arrayidx
+; AFTEREARLY:  polly.stmt.body:
```
Comment (nikic) [inline polly/test/Support/dumpmodule.ll:71]: Can you explain what actually changed in this test? It's hard to understand from this diff.

File: llvm/test/Transforms/InstCombine/array.ll (hunk 7 above)
Comment (LiqinWeng) [reply llvm/test/Transforms/InstCombine/array.ll:82]: done

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp (hunk 6 above)
Comment (LiqinWeng) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp]: done

**Comment by LiqinWeng** [comment]
friendly ping @nikic

File: llvm/test/Transforms/InstCombine/array.ll [hunk 9]
```diff
@@ -0,0 +1,82 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 3
+; RUN: opt < %s -passes=instcombine -S | FileCheck %s
+
+define void @test(ptr noundef %ptr, i32 noundef signext %a, i32 noundef signext %b) {
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/array.ll:4]: Drop noundef / signext.

File: llvm/test/Transforms/InstCombine/array.ll [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+; CHECK-NEXT:    [[TMP0:%.*]] = sext i32 [[A]] to i64
+; CHECK-NEXT:    [[TMP1:%.*]] = getelementptr i32, ptr [[PTR]], i64 [[TMP0]]
+; CHECK-NEXT:    [[GEP1:%.*]] = getelementptr i32, ptr [[TMP1]], i64 35
+; CHECK-NEXT:    store i32 [[ADD]], ptr [[GEP1]], align 4
+; CHECK-NEXT:    ret void
+;
+entry:
+  %add = add nsw i32 %a, 5
+  %idx = sext i32 %add to i64
+  %gep= getelementptr inbounds i32, ptr %ptr, i64 %idx
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/array.ll:39]:
```suggestion
  %gep = getelementptr inbounds i32, ptr %ptr, i64 %idx
```

File: llvm/test/Transforms/InstCombine/array.ll [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  store i32 %b, ptr %gep
+  %add1 = add nsw i32 %add, 30
+  %idx1 = sext i32 %add1 to i64
+  %gep1 = getelementptr inbounds i32, ptr %ptr, i64 %idx1
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/array.ll:43]: As far as I can tell, add1, idx1, gep1 are not really relevant to this test?

**Review by nikic — COMMENTED** [review]
This looks fine to me. @topperc Any concerns about this? I think you reported some RISCV regressions for the main add->gep change, so this might make things worse...

File: llvm/test/Transforms/InstCombine/array.ll (hunk 11 above)
Comment (LiqinWeng) [reply llvm/test/Transforms/InstCombine/array.ll:43]: It's really useless, I just want to reuse %add

File: llvm/test/Transforms/InstCombine/array.ll (hunk 9 above)
Comment (LiqinWeng) [reply llvm/test/Transforms/InstCombine/array.ll]: done

File: llvm/test/Transforms/InstCombine/array.ll (hunk 10 above)
Comment (LiqinWeng) [reply llvm/test/Transforms/InstCombine/array.ll]: done

File: llvm/test/Transforms/InstCombine/array.ll (hunk 11 above)
Comment (LiqinWeng) [reply llvm/test/Transforms/InstCombine/array.ll]: done

**Comment by LiqinWeng** [comment]
friendly ping O(∩_∩)O @nikic

**Review by nikic — APPROVED** [review]
LGTM

**Comment by brunodf-snps** [comment]
We observe that the `add + gep -> gep + gep` rewriting introduced in [D155688](https://reviews.llvm.org/D155688) (by @d-smirnov and @paulwalker-arm) and continued in #69581 here, breaks the LoopFlatten pass on loops with the following pattern [from coremark](https://github.com/eembc/coremark/blob/main/core_matrix.c#L285):

```
for (int i = 0; i < n; i++)
  for (int j = 0; j < n; j++)
     ... read/write A[i*n + j] ...
```

Godbolt link: https://godbolt.org/z/j3dKqj757 where clang/LLVM 18.1.0 has no loop flattening on 32 bit (bottom right output).

(See [similar report](https://github.com/llvm/llvm-project/issues/78214#issuecomment-2053804846) from @DragonDisciple which explains that the rewrite requires dropping `inbounds` from the original gep which blocks the optimization of the loop.)

It would seem from [bug 40581](https://bugs.llvm.org/show_bug.cgi?id=40581) and from [test pr40581.ll](https://github.com/llvm/llvm-project/blob/main/llvm/test/Transforms/LoopFlatten/pr40581.ll) that @LebedevRI @sjoerdmeijer explicitly intended LoopFlatten to support this loop pattern, but unfortunately, there is no test with a complete optimization pipeline for such a loop, except for [llvm/test/Transforms/PhaseOrdering/AArch64/loopflatten.ll](https://github.com/llvm/llvm-project/blob/main/llvm/test/Transforms/PhaseOrdering/AArch64/loopflatten.ll).

But since the latter test is only for a 64 bit target (where there is a `sext` to convert an `int` to the pointer index type), it did not fire at the time of D155688. Based on the discussion above, it seemed that the test did fire here, but it was fixed (inadvertently?) when @nikic [requested](https://github.com/llvm/llvm-project/pull/69581#pullrequestreview-1689395163) that the rewriting here be limited to the case where the RHS of the add is a constant (so the sext folds away).

That explains why loop flattening is not broken on 64 bit in the godbolt example above (`j` is not constant so the rewrite does not trigger), but D155688 did not have this restriction, and this gives rise to the situation that `add + gep` is only canonicalized to `gep + gep` based on a complex combination of conditions. Or it is at least very hard to explain under what conditions LLVM can still perform loop flattening of the above loop.

**Comment by nikic** [comment]
@brunodf-snps Assuming this is also an inbounds preservation problem as in the other reports, I think it should be possible to retain it in this case based on https://alive2.llvm.org/ce/z/wVFHGk. As far as I can see we do know that the add operands are non-negative here. I'll take a look at this.

**Comment by brunodf-snps** [comment]
@nikic Thanks! When the geps would retain the inbounds property, the loop flatten pass would indeed succeed:
https://godbolt.org/z/K64q46ah1
I shortly tried your patch with my C loop as input, but did not get it to retain the inbounds, even when the loop variables are `unsigned`. What is a scenario where you expect `isKnownNonNegative` to be derived?

**Comment by nikic** [comment]
Patch for the inbounds preservation: https://github.com/llvm/llvm-project/pull/90160

> @nikic Thanks! When the geps would retain the inbounds property, the loop flatten pass would indeed succeed: https://godbolt.org/z/K64q46ah1 I shortly tried your patch with my C loop as input, but did not get it to retain the inbounds, even when the loop variables are `unsigned`. What is a scenario where you expect `isKnownNonNegative` to be derived?

The problem here is that the IR at the time of the first InstCombine run is in non-rotated form (https://gist.github.com/nikic/e634b018bf63ae01c648516326f89faa). Lateron, we can very easily see that `%n` is non-negative, because there is an explicit condition for it in IR. This early, it's theoretically possible to derive this, but likely isn't feasible for compile-time reasons. So this may be a dead end for this example.

Worth noting that since https://github.com/llvm/llvm-project/pull/78576 LoopFlatten will still apply without inbounds via loop versioning, so this is not a question of whether flattening occurs, just of whether it requires versioning or not.

**Comment by brunodf-snps** [comment]
@nikic Thanks for the clarification. If I first perform loop rotation (https://gist.github.com/brunodf-snps/4a045f596a4c568fdeda86f691d9d118), I indeed see that inbounds is retained when instcombine rewrites the gep.

> Worth noting that since https://github.com/llvm/llvm-project/pull/78576 LoopFlatten will still apply without inbounds via loop versioning, so this is not a question of whether flattening occurs, just of whether it requires versioning or not.

Thanks for pointing this out. The loop versioning was added after LLVM 18 (see https://godbolt.org/z/xdYdYs3nn) so I had not noticed it yet.

**Pull request merged**